opm/simulators/wells/GasLiftStage2_impl.hpp

```diff
@@ -574,17 +573,41 @@ redistributeALQ_(std::vector<GasLiftSingleWell *> &wells,  const Opm::Group &gro
     std::vector<GradPair> &inc_grads, std::vector<GradPair> &dec_grads)
 {
     OptimizeState state {*this, group};
-    inc_grads.reserve(wells.size());
-    dec_grads.reserve(wells.size());
-    state.calculateEcoGradients(wells, inc_grads, dec_grads);
+    std::vector<GradPair> inc_grads_local;
+    std::vector<GradPair> dec_grads_local;
+    inc_grads_local.reserve(wells.size());
+    dec_grads_local.reserve(wells.size());
+
+    state.calculateEcoGradients(wells, inc_grads_local, dec_grads_local);
+
+    const auto& comm = ebos_simulator_.vanguard().grid().comm();
+    std::vector<int> sizes_(comm.size());
+    std::vector<int> displ_(comm.size() + 1, 0);
+    int mySize = inc_grads_local.size();
+    comm.allgather(&mySize, 1, sizes_.data());
+    std::partial_sum(sizes_.begin(), sizes_.end(), displ_.begin()+1);
+
+    std::vector<int> sizes2_(comm.size());
+    std::vector<int> displ2_(comm.size() + 1, 0);
+    int mySize2 = dec_grads_local.size();
+    comm.allgather(&mySize2, 1, sizes2_.data());
+    std::partial_sum(sizes2_.begin(), sizes2_.end(), displ2_.begin()+1);
+
+    std::vector<GradPair> inc_grads_global(displ_.back());
+    std::vector<GradPair> dec_grads_global(displ2_.back());
+
+    comm.allgatherv(inc_grads_local.data(), inc_grads_local.size(), inc_grads_global.data(), sizes_.data(), displ_.data());
+    comm.allgatherv(dec_grads_local.data(), dec_grads_local.size(), dec_grads_global.data(), sizes2_.data(), displ2_.data());
```

Comment:
GradPair is not POD, it's a pair of a string and a double.  Unless the communicator is really magic, this can't possibly work?

Reply:
Your probably right. I asked @blattms about this and got the impression that is should work, but I probably misunderstood. 
But I noticed that this code seems to work. 

https://github.com/OPM/opm-simulators/blob/3d738e83837f39bce036a22835203623fdb8dc95/opm/simulators/wells/ParallelWellInfo.cpp#L57-L75

I can try to communicate an index + double  pair instead and get the name from the index.

Reply:
that should work just fine. but data with variable size / pointer members won't. you can't define a MPI data type for variable sized data.

Reply:
Communicating a pair of index and double work. Thanks for help. It is not ready yet for other reasons. But the issue I had is fixed. I will continue on Monday....